rcl/test/rcl/common.cpp

```diff
@@ -0,0 +1,237 @@
+      RCUTILS_LOG_ERROR_NAMED(
+        ROS_PACKAGE_NAME, "Error in wait_set_add_client: %s", rcl_get_error_string().str);
+      return false;
+    }
+    ret = rcl_wait(&wait_set, RCL_MS_TO_NS(period_ms));
+    if (ret == RCL_RET_TIMEOUT) {
+      continue;
+    }
+    if (ret != RCL_RET_OK) {
+      RCUTILS_LOG_ERROR_NAMED(ROS_PACKAGE_NAME, "Error in wait: %s", rcl_get_error_string().str);
+      return false;
+    }
+    for (size_t i = 0; i < wait_set.size_of_clients; ++i) {
+      if (wait_set.clients[i] && wait_set.clients[i] == client) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+bool
+wait_for_service_to_be_ready(
+  rcl_service_t * service,
+  rcl_context_t * context,
+  size_t max_tries,
+  int64_t period_ms)
+{
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t ret =
+    rcl_wait_set_init(&wait_set, 0, 0, 0, 0, 1, 0, context, rcl_get_default_allocator());
+  if (ret != RCL_RET_OK) {
+    RCUTILS_LOG_ERROR_NAMED(
+      ROS_PACKAGE_NAME, "Error in wait set init: %s", rcl_get_error_string().str);
+    return false;
+  }
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    if (rcl_wait_set_fini(&wait_set) != RCL_RET_OK) {
+      RCUTILS_LOG_ERROR_NAMED(
+        ROS_PACKAGE_NAME, "Error in wait set fini: %s", rcl_get_error_string().str);
+      throw std::runtime_error("error waiting for service to be ready");
+    }
+  });
+  size_t iteration = 0;
+  while (iteration < max_tries) {
+    ++iteration;
+    if (rcl_wait_set_clear(&wait_set) != RCL_RET_OK) {
+      RCUTILS_LOG_ERROR_NAMED(
+        ROS_PACKAGE_NAME, "Error in wait_set_clear: %s", rcl_get_error_string().str);
+      return false;
+    }
+    if (rcl_wait_set_add_service(&wait_set, service, NULL) != RCL_RET_OK) {
+      RCUTILS_LOG_ERROR_NAMED(
+        ROS_PACKAGE_NAME, "Error in wait_set_add_service: %s", rcl_get_error_string().str);
+      return false;
+    }
+    ret = rcl_wait(&wait_set, RCL_MS_TO_NS(period_ms));
+    if (ret == RCL_RET_TIMEOUT) {
+      continue;
+    }
+    if (ret != RCL_RET_OK) {
+      RCUTILS_LOG_ERROR_NAMED(ROS_PACKAGE_NAME, "Error in wait: %s", rcl_get_error_string().str);
+      return false;
+    }
+    for (size_t i = 0; i < wait_set.size_of_services; ++i) {
+      if (wait_set.services[i] && wait_set.services[i] == service) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+bool
+wait_for_established_subscription(
+  const rcl_publisher_t * publisher,
+  size_t max_tries,
+  int64_t period_ms)
+{
+  size_t iteration = 0;
+  rcl_ret_t ret = RCL_RET_OK;
+  size_t subscription_count = 0;
+  while (iteration < max_tries) {
+    ++iteration;
+    ret = rcl_publisher_get_subscription_count(publisher, &subscription_count);
+    if (ret != RCL_RET_OK) {
+      RCUTILS_LOG_ERROR_NAMED(
+        ROS_PACKAGE_NAME,
+        "Error in rcl_publisher_get_subscription_count: %s",
+        rcl_get_error_string().str);
+      return false;
+    }
+    if (subscription_count == 1) {
```

Comment:
@Barry-Xu-2018 consider 
```suggestion
    if (subscription_count > 0) {
```
instead.

Reply:
Okay.